core/src/main/scala/filodb.core/query/QueryContext.scala

```diff
@@ -48,6 +48,82 @@ object PerQueryLimits {
   }
 
 }
+
+object QueryWarnings {
+
+}
+case class QueryWarnings(
+  execPlanSamples: AtomicInteger = new AtomicInteger(0),
+  execPlanResultBytes: AtomicLong = new AtomicLong(0),
+  groupByCardinality: AtomicInteger = new AtomicInteger(0),
+  joinQueryCardinality: AtomicInteger = new AtomicInteger(0),
+  timeSeriesSamplesScannedBytes: AtomicLong = new AtomicLong(0),
+  timeSeriesScanned: AtomicInteger = new AtomicInteger(0)
+) {
+
+  def hasWarnings() : Boolean = {
+    execPlanSamples.get() > 0 ||
+    execPlanResultBytes.get() > 0 ||
+    groupByCardinality.get() > 0 ||
+    joinQueryCardinality.get() > 0 ||
+    timeSeriesSamplesScannedBytes.get() > 0 ||
+    timeSeriesScanned.get() > 0
+  }
+
+  def merge(warnings: QueryWarnings) : Unit = {
+    updateExecPlanSamples(warnings.execPlanSamples.get())
+    updateExecPlanResultBytes(warnings.execPlanResultBytes.get())
+    updateGroupByCardinality(warnings.groupByCardinality.get())
+    updateJoinQueryCardinality(warnings.joinQueryCardinality.get())
+    updateTimeSeriesSampleScannedBytes(warnings.timeSeriesSamplesScannedBytes.get())
+    updateTimeSeriesScanned(warnings.timeSeriesScanned.get())
+  }
+
+  def updateExecPlanSamples(samples: Int): Unit = {
+    execPlanSamples.updateAndGet(s => if (s < samples) samples else s)
+  }
+
+  def updateExecPlanResultBytes(bytes: Long): Unit = {
+    execPlanResultBytes.updateAndGet(b => if (b < bytes) bytes else b)
+  }
+
+  def updateGroupByCardinality(cardinality: Int): Unit = {
+    groupByCardinality.updateAndGet(c => if (c < cardinality) cardinality else c)
+  }
+
+  def updateJoinQueryCardinality(cardinality: Int): Unit = {
+    joinQueryCardinality.updateAndGet(c => if (c < cardinality) cardinality else c)
+  }
+
+  def updateTimeSeriesScanned(series: Int): Unit = {
+    timeSeriesScanned.updateAndGet(s => if (s<series) series else s)
+  }
+
+  def updateTimeSeriesSampleScannedBytes(bytes: Long): Unit = {
+    timeSeriesSamplesScannedBytes.updateAndGet(b => if (b<bytes) bytes else b)
+  }
+
+  override def equals(w2Compare: Any): Boolean = {
+    val w2 = w2Compare.asInstanceOf[QueryWarnings]
```

Comment:
should you return false if w2Compare is other type

Reply:
it would throw an exception (which I thought is perfectly fine) but checked the equals implementation in java and decided to do the same thing, ie return false